scripts/oneoff/migrate-supplier-organisation-size-from-declarations.py

```diff
@@ -0,0 +1,128 @@
+    (HTTPTemporaryError, requests.exceptions.ConnectionError, RuntimeError),
+    max_tries=5,
+)
+
+
+def _catch_404_none(func):
+    try:
+        return func()
+    except HTTPError as e:
+        if e.status_code == 404:
+            return None
+        else:
+            raise
+
+
+def _get_supplier_frameworks(data_api_client, supplier_id):
+    # Get all supplierFrameworks for this supplier that have a declaration
+    try:
+        supplier_frameworks = _catch_404_none(
+            _backoff_wrap(
+                lambda: data_api_client.get_supplier_frameworks(supplier_id)
+            )
+        )
+    except StopIteration:
+        logger.info("Supplier %s: not on any relevant frameworks", supplier_id)
+        return None
+
+    # Filter out frameworks that do not have a declaration (i.e. containing the organisationSize)
+    return [sfw for sfw in supplier_frameworks.get('frameworkInterest', []) if sfw.get('declaration')]
+
+
+def _update_supplier_organisation_size(data_api_client, org_size, supplier_id, user, dry_run):
+    try:
+        logger.info("Updating supplier {} with org size {}".format(supplier_id, org_size))
+        if not dry_run:
+            _backoff_wrap(
+                lambda: data_api_client.update_supplier(supplier_id, {'organisationSize': org_size}, user=user)
+            )()
+    except HTTPError:
+        logger.info("HTTP ERROR UPDATING SUPPLIER {}".format(supplier_id))
+
+
+if __name__ == '__main__':
+    arguments = docopt(__doc__)
+
+    configure_logger({"script": loglevel_INFO})
+
+    client = DataAPIClient(get_api_endpoint_from_stage(arguments['<stage>']), arguments['<data_api_token>'])
+    user = arguments['<user>'] or getpass.getuser()
+    dry_run = bool(arguments.get("--dry-run"))
+    updated_suppliers_count = 0
+    start_time = datetime.utcnow()
+
+    for supplier in client.find_suppliers_iter():
+        logger.info("Processing supplier %s", supplier["id"])
+        if supplier.get("organisationSize"):
+            logger.info("  already done: {}".format(supplier["id"]))
+            continue
+
+        supplier_frameworks = _get_supplier_frameworks(client, supplier['id'])
+
+        if supplier_frameworks:
+            if len(supplier_frameworks) == 1:
+                supplier_framework = supplier_frameworks[0]
+            else:
+                # Get most recent framework declaration
+                supplier_framework = sorted(
+                    supplier_frameworks, key=lambda x: str(x['agreementReturnedAt']), reverse=True
+                )[0]
+
+            logger.info(
+                "Supplier %s: updating with data from framework %s", supplier["id"], supplier_framework["frameworkSlug"]
+            )
+
+            # Get the organisation size
+            org_size = supplier_framework.get('declaration', {}).get('organisationSize')
+            logger.info('  Updating with org size {}'.format(org_size))
+
+            if org_size in ['micro', 'small', 'medium', 'large']:
```

Comment:
If you're gonna test, you might as well output a warning if it doesn't match? (Ignoring `None` I suppose, that case isn't very interesting.)

Reply:
`None` would be interesting, because it's a required question in declarations, so shouldn't be any with `None`.

Reply:
Was it always required, or will there be people on DOS1 but not DOS2 that didn't fill it out perhaps (or G-Cloud 7 or something maybe?)